compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+                //     void foo(struct type1* bar) {}
+                //
+                // is encoded as:
+                //
+                //     !1 = !{i64 0, !"_ZTSFvP5type1E"}
+                //
+                // So, encode any repr(C) user-defined type [or extern types (i.e., ty::Foreign)] as
+                // <length><name>, where <name> is <unscoped-name>.
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                // Generalize any repr(transparent) user-defined type that is either a pointer or
+                // reference, and either references itself or any other type that contains or
+                // references itself, to avoid a reference cycle.
+                if ty0.is_any_ptr() && ty0.contains(ty) {
+                    s.push_str(&encode_ty(
+                        tcx,
+                        ty0,
+                        dict,
+                        options | EncodeTyOptions::GENERALIZE_POINTERS,
+                    ));
+                } else {
+                    s.push_str(&encode_ty(tcx, ty0, dict, options));
+                }
+            } else {
+                // u<length><name>[I<element-type1..element-typeN>E] as vendor extended type
+                s.push_str(&encode_ty_name(tcx, def_id));
+                if adt_def.is_struct() || adt_def.is_union() {
+                    let tys: Vec<Ty<'tcx>> = substs.types().map(|ty| ty).collect();
+                    if !tys.is_empty() {
+                        s.push('I');
+                        for ty in tys {
+                            s.push_str(&encode_ty(tcx, ty, dict, options));
+                        }
+                        s.push('E');
+                    }
+                }
+                s = format!("u{}{}", s.len(), &s);
+                compress(dict, (ty, TyQ::None), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::Foreign(def_id) => {
+            // <length><name>, where <name> is <unscoped-name>
+            let mut s = String::new();
+            let name = format!("{}", tcx.item_name(*def_id));
+            s.push_str(&format!("{}", name.len()));
+            s.push_str(&name);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Function types
+        ty::FnDef(def_id, ..) => {
+            // u2fnI<length><name>E as vendor extended type
+            let mut s = String::from("u2fnI");
+            s.push_str(&encode_ty_name(tcx, *def_id));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
```

Comment:
This should encode both `def_id` and `substs`.

Reply:
What do you mean by encode `def_id`?

Reply:
I mean mangle the path corresponding to `def_id`, just it is already done for `ty::Adt` with `encode_ty_name`.

Reply:
It's already being done for this `ty::Ty` variant with `encode_ty_name`. Am I missing something?

Reply:
It also needs to encode the `substs` in the case function is generic.

Reply:
Yes, I'll add the `substs`.

Reply:
Done.